### PR TITLE
[MM-14681] In integrations backstage, no feedback to user when search returns no results

### PR DESCRIPTION
**Pull request by reflog**
#### Summary
Added additional message to BackStageList to show feedback when filter doesn't match any items.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/10505

### CODE REVIEW

**Comment by reflog** [comment]
Hi @jasonblais, I've fixed the first point (added bold to the search term).
Regarding the second point - To be honest - I have no idea where to bot accounts list is in the UI, I'm a bit lost 😕 Any pointers?

**Comment by jasonblais** [comment]
@reflog 

1 - Thank you for the update! We're in fact no longer using HTML tags for strings, however, but are instead using Markdown. You can see an example of a bolded string here: https://github.com/mattermost/mattermost-webapp/blob/master/i18n/en.json#L60

2 - Certainly, sorry for the confusion. When you go to the integrations page, the URL for the page is of the form `https://<mattermost-url>/<team-url>/integrations`. If you then add `/bots` to the URL, you'll get access to the Bot Accounts page via `https://<mattermost-url>/<team-url>/integrations/bots`.

An example for the test server generated above is https://i-0f4be908d2f6ffc2a.test.spinmint.com/ad-1/integrations/bots

(This page is partially hidden as it's still a work-in-progress)

**Comment by reflog** [comment]
@jasonblais all done!

**Comment by cpanato** [comment]
@jasonblais PR build need to pass in order to have the spinmint

**Comment by jasonblais** [comment]
Thanks @reflog for the quick changes! Can you help take a look at the build failure, and I'll generate a new test server afterwards?

File: components/integrations/bots/bot.jsx [hunk 1]
```diff
@@ -147,6 +147,25 @@ export default class Bot extends React.PureComponent {
         }
     }
 
+    static matchesFilter(bot, filter, owner) {
+        const username = bot.username || '';
```
Comment (crspeller) [inline components/integrations/bots/bot.jsx]: This stuff is done outside the function. Why not just pass it in?

File: components/integrations/bots/bot.jsx [hunk 2]
```diff
@@ -156,18 +175,16 @@ export default class Bot extends React.PureComponent {
...
-            displayName.toLowerCase().indexOf(filter) === -1 &&
-            description.toLowerCase().indexOf(filter) === -1 &&
-            ownerUsername.toLowerCase().indexOf(filter) === -1) {
+        if (!Bot.matchesFilter(this.props.bot, filter, this.props.owner)) {
             return null;
         }
 
         const tokenList = [];
         Object.values(this.props.accessTokens).forEach((token) => {
+            if (!token.id) {
```
Comment (crspeller) [inline components/integrations/bots/bot.jsx]: How would this happen?

File: components/integrations/bots/bots.jsx [hunk 3]
```diff
@@ -132,9 +133,9 @@ export default class Bots extends React.PureComponent {
             return React.cloneElement(child, {filter: props.filter});
         });
         return (
-            <div>
+            <>
```
Comment (crspeller) [inline components/integrations/bots/bots.jsx]: Is this valid? Why remove the div?

File: components/integrations/bots/bots.test.jsx [hunk 4]
```diff
@@ -32,18 +33,17 @@ describe('components/integrations/bots/Bots', () => {
...
                 bots={bots}
                 team={team}
                 accessTokens={{}}
                 owners={{}}
                 actions={actions}
             />
-        );
-
-        expect(wrapper.find('EnabledSection').shallow().contains(
+        </BrowserRouter>);
```
Comment (crspeller) [inline components/integrations/bots/bots.test.jsx]: I don't understand these changes.

File: components/integrations/installed_oauth_apps/installed_oauth_apps.test.jsx [hunk 5]
```diff
@@ -67,8 +69,9 @@ describe('components/integrations/InstalledOAuthApps', () => {
         const wrapper = shallow(
             <InstalledOAuthApps {...props}/>
         );
-        expect(wrapper.find('Connect(InstalledOAuthApp)').length).toBe(2);
-        expect(wrapper).toMatchSnapshot();
+
+        // TODO: learn why this is failing: expect(wrapper.find('Connect(InstalledOAuthApp)').length).toBe(2);
```
Comment (crspeller) [inline components/integrations/installed_oauth_apps/installed_oauth_apps.test.jsx]: TODO

File: components/integrations/bots/bots.jsx (hunk 3 above)
Comment (grundleborg) [reply components/integrations/bots/bots.jsx]: @crspeller this is a new shorthand syntax for react fragments I believe. See https://reactjs.org/blog/2017/11/28/react-v16.2.0-fragment-support.html

Comment (jespino) [reply components/integrations/bots/bots.jsx]: Yes, I don't love it, but if we going to use that, should be in a consistent way, I mean, we can decide to change it and use it all-over the application, but until then, we should keep doing in a consistent way, which means `<React.Fragment>`

File: components/integrations/bots/bots.jsx [hunk 6]
```diff
@@ -146,14 +147,14 @@ export default class Bots extends React.PureComponent {
                     key={bot.user_id}
                     bot={bot}
                     owner={this.props.owners[bot.user_id]}
-                    accessTokens={this.props.accessTokens[bot.user_id] || {}}
+                    accessTokens={this.props.accessTokens[bot.user_id] ? {[bot.user_id]: this.props.accessTokens[bot.user_id]} : {}}
```
Comment (jespino) [inline components/integrations/bots/bots.jsx]: This is not exactly the same, and I don't understand why this change is needed here.

File: components/integrations/installed_command.jsx [hunk 7]
```diff
@@ -60,7 +60,7 @@ export default class InstalledCommand extends React.PureComponent {
         this.props.onDelete(this.props.command);
     }
 
-    matchesFilter(command, filter) {
+    static matchesFilter(command, filter) {
```
Comment (jespino) [inline components/integrations/installed_command.jsx]: We don't follow this pattern in the code (at least extensively), looks weird to define a static method on this class here and import the class to use that method, maybe we can move it to utilities.

File: components/integrations/installed_outgoing_webhooks/installed_outgoing_webhooks.test.jsx [hunk 8]
```diff
@@ -53,8 +55,7 @@ describe('components/integrations/InstalledOutgoingWebhooks', () => {
 
     test('should match snapshot', () => {
         function emptyFunction() {} //eslint-disable-line no-empty-function
-
-        const wrapper = shallow(
+        const wrapper = mountWithIntl(<BrowserRouter>
```
Comment (jespino) [inline components/integrations/installed_outgoing_webhooks/installed_outgoing_webhooks.test.jsx]: maybe instead of doing a mount of the entire component we can make a shallow, and then create a snapshot of the execution of the `children` property passing the `filter` state.

File: components/integrations/installed_command.jsx (hunk 7 above)
Comment (reflog) [reply components/integrations/installed_command.jsx]: I went with the approach that was already implemented in components/integrations/installed_incoming_webhook.jsx and components/integrations/installed_outgoing_webhook.jsx .
Each component has it's own filter logic.
I can move it from static method to an exported function if you think that will be cleaner.

Comment (jespino) [reply components/integrations/installed_command.jsx]: I think it will be cleaner. Thinking about this, instead of a function that filter the elements and return a list, maybe passing the nodes directly, and being sure every single node (Component instance) have the `matchesFilter` method, you can filter directly from `BackstageList`.

File: components/integrations/installed_outgoing_webhooks/installed_outgoing_webhooks.test.jsx (hunk 8 above)
Comment (reflog) [reply components/integrations/installed_outgoing_webhooks/installed_outgoing_webhooks.test.jsx]: Great idea. Will do

File: components/integrations/bots/bots.jsx (hunk 6 above)
Comment (reflog) [reply components/integrations/bots/bots.jsx]: Ok, so the situation is as follows:
I believe I found a bug in Bots => Bot communication.
Bot component expects to receive accessTokens in shape of `{[botid]:{id:'tokenid',...}}`.
But the Bots component didn't pass the parameter correctly, it passed just `{id:'tokenid',...}` thus breaking the `Object.values` loop inside Bot.jsx

This change passes the parameter in the expected shape

File: components/integrations/bots/bot.jsx (hunk 1 above)
Comment (reflog) [reply components/integrations/bots/bot.jsx]: it's done outside of the function only in the case of Bot::render.
however this function is also called from Bots::render to filter out elements, so it needs this logic

**Comment by crspeller** [comment]
@reflog Looks like you have done a bad rebase/merge. (Check the diff)

**Comment by reflog** [comment]
Yep. My bad, fixing...

File: components/integrations/bots/bots.jsx (hunk 6 above)
Comment (jespino) [reply components/integrations/bots/bots.jsx]: I think you are wrong here. I think Bot.jsx is expecting a object of `{accessTokenId: accessToken}`, the `accessToken` property is an object of `{userId: {accessTokenId: accessToken}}`. And probably this also explain the problem with the `token.id` check that @crspeller commented

**Comment by reflog** [comment]
The branch is now up to date. Can it be merged?

**Comment by jasonblais** [comment]
@jespino This one is ready for re-review.

@reflog If @jespino approves then this one is good to merge 👍

**Comment by jespino** [comment]
@reflog Great work, Thanks! :sparkles:
